File: java/lib/src/main/java/com/svix/models/SnowflakeConfig.java
// This file is @generated
package com.svix.models;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.svix.Utils;

import lombok.EqualsAndHashCode;
import lombok.ToString;

@ToString
@EqualsAndHashCode
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonAutoDetect(getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class SnowflakeConfig {
    @JsonProperty private String accountIdentifier;
    @JsonProperty private String dbName;
    @JsonProperty private String privateKey;
    @JsonProperty private String schemaName;
    @JsonProperty private String tableName;
    @JsonProperty private String userId;

    public SnowflakeConfig() {}

    public SnowflakeConfig accountIdentifier(String accountIdentifier) {
        this.accountIdentifier = accountIdentifier;
        return this;
    }

    /**
     * Snowflake account identifier, which includes both the organization and account IDs separated
     * by a hyphen.
     *
     * @return accountIdentifier
     */
    @javax.annotation.Nonnull
    public String getAccountIdentifier() {
        return accountIdentifier;
    }

    public void setAccountIdentifier(String accountIdentifier) {
        this.accountIdentifier = accountIdentifier;
    }

    public SnowflakeConfig dbName(String dbName) {
        this.dbName = dbName;
        return this;
    }

    /**
     * Database name.
     *
     * <p>Only required if not using transformations.
     *
     * @return dbName
     */
    @javax.annotation.Nullable
    public String getDbName() {
        return dbName;
    }

    public void setDbName(String dbName) {
        this.dbName = dbName;
    }

    public SnowflakeConfig privateKey(String privateKey) {
        this.privateKey = privateKey;
        return this;
    }

    /**
     * PEM-encoded private key used for signing token-based requests to the Snowflake API.
     *
     * <p>Beginning/end delimiters are not required.
     *
     * @return privateKey
     */
    @javax.annotation.Nonnull
    public String getPrivateKey() {
        return privateKey;
    }

    public void setPrivateKey(String privateKey) {
        this.privateKey = privateKey;
    }

    public SnowflakeConfig schemaName(String schemaName) {
        this.schemaName = schemaName;
        return this;
    }

    /**
     * Schema name.
     *
     * <p>Only required if not using transformations.
     *
     * @return schemaName
     */
    @javax.annotation.Nullable
    public String getSchemaName() {
        return schemaName;
    }

    public void setSchemaName(String schemaName) {
        this.schemaName = schemaName;
    }

    public SnowflakeConfig tableName(String tableName) {
        this.tableName = tableName;
        return this;
    }

    /**
     * Table name.
     *
     * <p>Only required if not using transformations.
     *
     * @return tableName
     */
    @javax.annotation.Nullable
    public String getTableName() {
        return tableName;
    }

    public void setTableName(String tableName) {
        this.tableName = tableName;
    }

    public SnowflakeConfig userId(String userId) {
        this.userId = userId;
        return this;
    }

    /**
     * The Snowflake user id.
     *
     * @return userId
     */
    @javax.annotation.Nonnull
    public String getUserId() {
        return userId;
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    /**
     * Create an instance of SnowflakeConfig given an JSON string
     *
     * @param jsonString JSON string
     * @return An instance of SnowflakeConfig
     * @throws JsonProcessingException if the JSON string is invalid with respect to SnowflakeConfig
     */
    public static SnowflakeConfig fromJson(String jsonString) throws JsonProcessingException {
        return Utils.getObjectMapper().readValue(jsonString, SnowflakeConfig.class);
    }

    /**
     * Convert an instance of SnowflakeConfig to an JSON string
     *
     * @return JSON string
     */
    public String toJson() throws JsonProcessingException {
        return Utils.getObjectMapper().writeValueAsString(this);
    }
}

File: php/src/Models/SnowflakeConfig.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Models;

/** Configuration parameters for defining a Snowflake sink. */
class SnowflakeConfig implements \JsonSerializable
{
    private array $setFields = [];

    /**
     * @param string $privateKey PEM-encoded private key used for signing token-based requests to the Snowflake API.
     *
     * Beginning/end delimiters are not required.
     * @param string      $accountIdentifier snowflake account identifier, which includes both the organization and account IDs separated by a hyphen
     * @param string      $userId            the Snowflake user id
     * @param string|null $dbName            Database name.
     *
     * Only required if not using transformations.
     * @param string|null $schemaName Schema name.
     *
     * Only required if not using transformations.
     * @param string|null $tableName Table name.
     *
     * Only required if not using transformations.
     */
    private function __construct(
        public readonly string $privateKey,
        public readonly string $accountIdentifier,
        public readonly string $userId,
        public readonly ?string $dbName = null,
        public readonly ?string $schemaName = null,
        public readonly ?string $tableName = null,
        array $setFields = [],
    ) {
        $this->setFields = $setFields;
    }

    /**
     * Create an instance of SnowflakeConfig with required fields.
     */
    public static function create(
        string $privateKey,
        string $accountIdentifier,
        string $userId,
    ): self {
        return new self(
            privateKey: $privateKey,
            accountIdentifier: $accountIdentifier,
            userId: $userId,
            dbName: null,
            schemaName: null,
            tableName: null,
            setFields: ['privateKey' => true, 'accountIdentifier' => true, 'userId' => true]
        );
    }

    public function withDbName(?string $dbName): self
    {
        $setFields = $this->setFields;
        $setFields['dbName'] = true;

        return new self(
            privateKey: $this->privateKey,
            accountIdentifier: $this->accountIdentifier,
            userId: $this->userId,
            dbName: $dbName,
            schemaName: $this->schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withSchemaName(?string $schemaName): self
    {
        $setFields = $this->setFields;
        $setFields['schemaName'] = true;

        return new self(
            privateKey: $this->privateKey,
            accountIdentifier: $this->accountIdentifier,
            userId: $this->userId,
            dbName: $this->dbName,
            schemaName: $schemaName,
            tableName: $this->tableName,
            setFields: $setFields
        );
    }

    public function withTableName(?string $tableName): self
    {
        $setFields = $this->setFields;
        $setFields['tableName'] = true;

        return new self(
            privateKey: $this->privateKey,
            accountIdentifier: $this->accountIdentifier,
            userId: $this->userId,
            dbName: $this->dbName,
            schemaName: $this->schemaName,
            tableName: $tableName,
            setFields: $setFields
        );
    }

    public function jsonSerialize(): mixed
    {
        $data = [
            'privateKey' => $this->privateKey,
            'accountIdentifier' => $this->accountIdentifier,
            'userId' => $this->userId];

        if (null !== $this->dbName) {
            $data['dbName'] = $this->dbName;
        }
        if (null !== $this->schemaName) {
            $data['schemaName'] = $this->schemaName;
        }
        if (null !== $this->tableName) {
            $data['tableName'] = $this->tableName;
        }

        return \Svix\Utils::newStdClassIfArrayIsEmpty($data);
    }

    /**
     * Create an instance from a mixed obj.
     */
    public static function fromMixed(mixed $data): self
    {
        return new self(
            privateKey: \Svix\Utils::deserializeString($data, 'privateKey', true, 'SnowflakeConfig'),
            accountIdentifier: \Svix\Utils::deserializeString($data, 'accountIdentifier', true, 'SnowflakeConfig'),
            userId: \Svix\Utils::deserializeString($data, 'userId', true, 'SnowflakeConfig'),
            dbName: \Svix\Utils::deserializeString($data, 'dbName', false, 'SnowflakeConfig'),
            schemaName: \Svix\Utils::deserializeString($data, 'schemaName', false, 'SnowflakeConfig'),
            tableName: \Svix\Utils::deserializeString($data, 'tableName', false, 'SnowflakeConfig')
        );
    }

    /**
     * Create an instance from a json string.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode(json: $json, associative: true, depth: 512, flags: JSON_THROW_ON_ERROR);

        return self::fromMixed($data);
    }
}
